ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskRequestSenderTest.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright contributors to Hyperledger Besu.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.hyperledger.besu.ethereum.eth.manager.peertask;
+
+import org.hyperledger.besu.ethereum.eth.manager.EthPeer;
+import org.hyperledger.besu.ethereum.eth.manager.RequestManager;
+import org.hyperledger.besu.ethereum.p2p.rlpx.connections.PeerConnection;
+import org.hyperledger.besu.ethereum.p2p.rlpx.wire.MessageData;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Mockito;
+
+public class PeerTaskRequestSenderTest {
+
+  private PeerTaskRequestSender peerTaskRequestSender;
+
+  @BeforeEach
+  public void beforeTest() {
+    peerTaskRequestSender = new PeerTaskRequestSender();
+  }
+
+  @Test
+  public void testSendRequest()
+      throws PeerConnection.PeerNotConnected, ExecutionException, InterruptedException {
+    String subprotocol = "subprotocol";
+    MessageData requestMessageData = Mockito.mock(MessageData.class);
+    MessageData responseMessageData = Mockito.mock(MessageData.class);
+    EthPeer peer = Mockito.mock(EthPeer.class);
+    PeerConnection peerConnection = Mockito.mock(PeerConnection.class);
+    RequestManager.ResponseStream responseStream =
+        Mockito.mock(RequestManager.ResponseStream.class);
+
+    Mockito.when(peer.getConnection()).thenReturn(peerConnection);
+    Mockito.when(peer.send(requestMessageData, subprotocol, peerConnection))
+        .thenReturn(responseStream);
+
+    CompletableFuture<MessageData> actualResponseMessageDataFuture =
+        CompletableFuture.supplyAsync(
+            () -> {
+              try {
+                return peerTaskRequestSender.sendRequest(subprotocol, requestMessageData, peer);
+              } catch (Exception e) {
+                throw new RuntimeException(e);
+              }
+            });
+    Thread.sleep(500);
```

Comment:
Why do we need this in the test?

Reply:
Because of the nature of RequestSender, the actual test has to be performed on another thread than the test thread. The code below this line must be execute after RequestSender has submitted it's callback method to responseStream.then. Since it doesn't make sense to expose a signal just for testing, we can just add a small delay to ensure the code is executed in the required sequence. Obviously, for production code, this is a dirty hack and proper inter-thread signalling should be used.